쥬스 메이커 [STEP1] Whales

WhalesJin opened the pull request:
안녕하세요 mmim(@joSH0318) :wolf: !!
9기 중 첫 개인 플젝을 하게 된 머쨍이 Whales🐋입니다!!
2주 동안 가르쳐주시는 모든 부분들 열심히 배워보겠습니다!!
잘 부탁드립니다아🙇🏻‍♀️

### 💡 구상에 대한 고민
> JuiceMaker, FruitStore 객체의 역할
- `JuiceMaker`는 쥬스 제조를 담당하는 객체이고, 고객의 주문을 받아 해당하는 쥬스를 만드는 역할을 합니다.특히, 필요한 과일의 종류와 수량을 파악하고, `FruitStore`에서 필요한 과일을 가져와 쥬스를 제조합니다. 만약 필요한 과일이 충분하지 않다면, 에러를 발생시킵니다.
- `FruitStore`는 과일 재고를 관리하는 객체이고, 과일의 재고를 증가시키거나 감소시키는 기능을 가지고 있습니다. 또한, 특정 과일의 현재 재고를 확인하는 기능도 제공합니다. `JuiceMaker`가 요청하는 과일을 제공하거나, 재고가 부족하다는 것을 알려주는 역할을 합니다.
> 이번 프로젝트에서의 Model의 역할
- 저희가 학습한 MVC 모델에 따르면 모델은 데이터와 비즈니스 로직을 처리하는 역할을 담당하는 구성 요소입니다. 즉, 데이터 저장, 비즈니스 로직 처리, 데이터 접근 및 조작을 관리하는 역할이라고 생각합니다. 이후 스텝에서도 데이터와 비즈니스 로직의 구현이 필요한 부분은 모두 모델의 역할이라고 생각합니다.
> method 배치 순서
- 저희는 호출 순서에 따라 메소드를 배치하였습니다. `private` 메소드는 하단에 배치하는 것 또한 좋을 것이라는 생각도 들었지만, 본 프로젝트에서는 호출 순서를 고려하여 배치하였습니다.

### 📝 고민한 점
> 큰 그림이 중요하다!
- 사실 아직은 앱을 만드는 부분이나 UI부분이 낯설어서 step을 진행하며 전처럼 요구사항만 읽고 로직을 짜는데 집중했습니다. 그러다보니 진행하면서 넣을까 말까 고민하는 기능들, 넣고나니 더 꼬이기만 하는 기능들, 충돌되던 컨셉들을 발견하며 탁탁 막히는 느낌을 받았습니다.
- 프로젝트를 진행하며 `Main.storyboard`라는 부분을 보게 되었는데 이제까지 하던 고민이 결국 이로 이어지는 고민들이라는 것을 깨닫게 되었습니다. 이미 구성되어 있는 화면을 보며 타입 안에 필요한 프로퍼티나 메서드 등을 고민하고 구현 과정을 계획했다면 갈팡질팡 했던 긴 고민의 시간이 줄었을 것 같다는 생각을 했고, 앞으로 그런 방향도 같이 고려해서 로직을 짠다면 좀 더 효율적이고 완성도가 높은 코드를 구현하는데 큰 도움이 될 것 같다는 깨달음을 얻었습니다.
- (스토리보드를 미리 봤다면 고민의 시간이 줄었을 수 있는 부분)을 📱 이모티콘으로 표시

> JuiceMaker - struct, FruitStore - class 사용 기준
- 처음에 코드를 새로 짤 때, 미션 파일에 `JuiceMaker`는 `struct`로, `FruitStore`는 `class`로 깔려있었는데 참조하거나 상속할 일이 없으니 class를 struct로 변경해서 진행했습니다.
- 완성 후 테스트를 돌려보는데
    ![](https://hackmd.io/_uploads/H1hMoEjN2.png)
    `qoocrab` 안에 `whales` 과일 가게가 있는데 `whales`에 있는 `fruitInventory`와 `qoocrab`에 있는 `fruitInventory`가 출력이 다르게 나오는 걸 보고 뭔가 잘못되었음을 깨달았습니다.
- 같은 재고를 가지고 있어야는데 `struct`를 써서 값 참조로 복사가 되었기 때문에 연동이 안됨을 찾아서 `FruitStore`를 `class`로 다시 바꿔서 해결했습니다.
- 프로젝트를 진행할 때마다 `class`와 `struct`에 대해서 좀 더 고민을 해보고 진행을 하면 좋겠다 라는 걸 직접 느끼게 되는 부분이었습니다.
- 📱

> static
- 체인점의 개념을 생각했을 때, 모두의 레시피가 동일해야하고 한 군데의 레시피가 변경되면 다같이 바꾸기 위해서 static이라는 키워드를 떠올려보았습니다.
- `static` 변수는 여러 곳에서 공유하며 참조의 형태로 한 곳에서 변경되면 다같이 변경되는 특성을 가지고 있습니다.
- 단점은 프로그램이 종료될 때까지 메모리에 남아있어서 과도하게 사용하면 불필요한 메모리 소모가 발생될 수 있고, 여려 곳에서 공유하기 때문에 의도치 않게 변경될 수 있습니다.
- 추상화의 중요성을 느꼈습니다. 지금 우리가 구현해야하는 부분과 상관없는 고민임을 깨닫고 구현하지는 않았지만 덕분에 새로운 키워드를 공부하게 되었습니다.
- 📱

> Hashable 프로토콜 준수
- `Juice`라는 타입을 만들며 프로퍼티로 쥬스의 레시피를 `Dictionary`로 구현하였습니다. `Dictionary`가 `key-value` 값을 저장하는 해시 테이블로 구현이 되어있어서 그 `key`로 쓰여지는 타입은 `Hashable` 프로토콜을 준수해야 합니다. 그래서 `Fruit` 타입은 `Hashable` 프로토콜을 준수하도록 만들었습니다.
- 연관값 없는 열거형은 명시해주지 않더라도 자동으로 `Hashable` 채택이 되어서 키워드를 삭제하였습니다.
> 네이밍
- 변수 네이밍 중 특히 `get`의 사용을 고민했다거나, 동사가 빠져있다거나 하는 경우가 많았었습니다. 그래서 이를 대체하는 동사를 고민을 많이 했습니다.
- `get`과 같은 예약어들은 사용처가 정해져 있기에 대체 단어가 있다면 피하는 게 좋고 make로 시작하는 `method`들은 보는 사람으로 하여금 `factory method`를 떠올리게 할 수 있어 피하는게 좋다는 걸 다시 배우고 상기하게 되었습니다.
- [API Design Guideline - Strive for Fluent Usage](https://www.swift.org/documentation/api-design-guidelines/#promote-clear-usage)

> 에러처리 구현
- 에러를 던지고 잡는 시점은 조금 알아가고 있는데 정작 `try` 했을 때 돌아가는 형태를 제대로 숙지하지 못해서 코드가 복잡해졌습니다.
    ```swift
    if !(try fruitStore.isStockSufficient(fruit, with: amount)) {
        return false
    }
    ```
    - `try`로 에러 던지는 함수를 호출하게 되면 실패시 자연스럽게 에러를 던지게 되므로 `if문` 같은 조건문이 필요하지 않다는 것을 배워서 아래와 같이 수정하였습니다.
        ```swift
        try fruitStore.isStockSufficient(fruit, with: amount)
        ``` 
- `isStockSufficient` 함수를 반환 타입을 `Bool` 값으로 구현하였는데, 최종적으로 `takeOrder`에서 에러처리를 해주고 있으니` Bool` 값으로 함수의 단계를 더 쪼개기보다는 `false`인 경우에 에러를 던지면 좋을 것 같다는 조언을 받고 수정하였습니다.
    - before
    ```swift
    func isStockSufficient(_ fruit: Fruit, with requiredAmount: Int) throws -> Bool {
            let currentStock = try readCurrentStock(for: fruit)
            return currentStock >= requiredAmount
        }
    ```
    - after
    ```swift
    func isStockSufficient(_ fruit: Fruit, with requiredAmount: Int) throws {
        let currentStock = try readCurrentStock(for: fruit)
        guard currentStock >= requiredAmount else {
            throw JuiceMakerError.insufficientFruit
        }
    }
    ```

WhalesJin committed:
feat: 과일 열거형 타입 생성

WhalesJin committed:
feat: 쥬스 열거형 타입 생성

- 과일쥬스 케이스
- name 프로퍼티 String으로
- recipe 프로퍼티 Dictionary로, [과일: 개수]

WhalesJin committed:
feat: 에러 열거형 타입 생성

WhalesJin committed:
feat: 과일가게 클래스 타입 생성

- 과일 재고품 프로퍼티
- readCurrentStock 함수
- isStockSufficient 함수
- decreaseFruitStock 함수
- increaseFruitStock 함수

WhalesJin committed:
feat: 쥬스메이커 구조체 타입 생성

- 과일 가게 프로퍼티 (재고 10)
- takeOrder 함수 (do-try-catch)
- makeJuice 함수
- checkStock 함수
- consumeFruit 함수

WhalesJin committed:
Create docs: README.md 작성

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,43 @@
+        try consumeFruit(for: juice)
+    }
+
+    private func checkStock(for juice: Juice) throws {
```
`check`는 검사하다라는 뜻이지만 
"옳다, 그르다"라는 느낌이라서 bool 값을 반환할 것이라 예상하게 됩니다.
수정하면 좋을 것 같아요!

WhalesJin replied:
좋은 조언 감사드립니다!!
`check`라는 이름을 `verify`로 수정하였습니다!!

JoSH0318 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,34 @@
+        return currentAmount
+    }
+    
+    func isStockSufficient(_ fruit: Fruit, with requiredAmount: Int) throws {
+        let currentStock = try readCurrentStock(for: fruit)
+        guard currentStock >= requiredAmount else {
+            throw JuiceMakerError.insufficientFruit
+        }
+    }
+    
+    func decreaseFruitStock(_ fruit: Fruit, amount: Int) throws {
+        let currentAmount = try readCurrentStock(for: fruit)
+        fruitInventory[fruit] = currentAmount - amount
+    }
```
두 메서드는 같은 파라미터를 받고있지만 paramter name이 다르네요? 🥹

`decreaseFruitStock` 에도 적당한 argument label을 정해주면 좋을 것 같아요.

WhalesJin replied:
`by` 레이블을 달아줬더니 호출할 때 훨씬 예뻐졌습니다!! 감사합니다!!

JoSH0318 reviewed (commented):
안녕하세요 웨일!
step1 첫번째 pr 이네요! 고생하셨습니다.
코드에 코멘트 남겼습니다. 확인하시고 수정 부탁드립니다! 
궁금한 점, 이해가 안되는 점, 이견이 있다면 언제든 말씀해주세요!

WhalesJin committed:
refactor: 네이밍 수정

-  decrease/increase FruitStock 함수의 amount 파라미터에 by 레이블 추가
- checkStock 함수를 verifyStock으로 이름 수정
- InvalidJuice 에러를 지우고 에러 타입 이름을 FruitStoreError로 수정

JoSH0318 commented:
고생하셨습니다 !
요구사항 만족하셨고 특별히 언급할 점이 없어 merge 하도록 하겠습니다.
step2도 화이팅하세요!
